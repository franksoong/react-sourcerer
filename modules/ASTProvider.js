import React from 'react';
import {parse} from 'babylon';

class ASTProvider extends React.Component {
  getChildContext() {
    return {
      ast: parse(this.props.source, {
        sourceType: 'module',
        plugins: [
          'jsx',
        ],
      }),
    };
  }

  render() {
    return (
      React.Children.only(this.props.children)
    );
  }
}

ASTProvider.childContextTypes = {
  ast: React.PropTypes.object,
};

export default ASTProvider;
