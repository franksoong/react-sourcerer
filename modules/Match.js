import React from 'react';
import traverse from 'babel-traverse';

class Match extends React.Component {
  render() {
    const matches = [];
    const matchTest = this.props.match;

    const walk = {
      enter(path) {
        const isMatch = matchTest(path);

        if (isMatch) {
          matches.push(
            path
          );
        }
      },
    };

    if (this.props.ast) {
      this.props.ast.traverse(walk);
    } else {
      traverse(this.context.ast, walk);
    }

    return React.Children.only(this.props.children(matches));
  }
}

Match.contextTypes = {
  ast: React.PropTypes.object,
};

export default Match;
