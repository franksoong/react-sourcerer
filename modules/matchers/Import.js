import React from 'react';
import Match from '../Match';
import * as t from 'babel-types';

class Import extends React.Component {
  render() {
    return (
      <Match
        {...this.props}
        match={(path) => {
          const {source} = this.props;
          const isImportDeclaration = t.isImportDeclaration(path);
          if (!isImportDeclaration) return false;
          if (source && source !== path.node.source.value) return false;
          return isImportDeclaration;
        }}
      />
    );
  }
}

export default Import;
