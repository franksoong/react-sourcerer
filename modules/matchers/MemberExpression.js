import React from 'react';
import Match from '../Match';
import * as t from 'babel-types';

class MemberExpression extends React.Component {
  handleMatch = (path) => {
    const { name, target } = this.props;
    const isMemberExpression = t.isMemberExpression(path);
    if (!isMemberExpression) return false;
    if (target && Array.isArray(target) && target.includes(path.node.object.name)) return true;
    if (target && target !== path.node.object.name) return false;
    if (name && name !== path.node.property.name) return false;
    return isMemberExpression;
  }

  render() {
    return (
      <Match
        {...this.props}
        match={this.handleMatch}
      />
    );
  }
}

export default MemberExpression;
