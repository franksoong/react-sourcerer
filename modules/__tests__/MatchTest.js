import ASTProvider from '../ASTProvider';
import React from 'react';
import Match  from '../Match';
import Import, {getIdentifiersFromImport} from '../matchers/Import';
import MemberExpression from '../matchers/MemberExpression';
import * as t from 'babel-types';

import renderer from 'react-test-renderer';


describe('Match', () => {
  it('should match a class method', () => {
    const output = renderer.create(<ASTProvider source={`
      class Person {
        greet() {
        }
      }
    `}>
      <Match
        match={(path) => {
          return t.isClassMethod(path);
        }}
      >
        {
          ((matches) => {
            return (
              <div>
                {matches.map((path) => path.node.type)}
              </div>
            );
          })
        }
      </Match>
    </ASTProvider>);

    expect(output.toJSON()).toMatchSnapshot(); 
  });

  it('should match a variable', () => {
    const output = renderer.create(<ASTProvider source={`
      const name = 'Merrick';
    `}>
      <Match
        match={(path) => {
          return t.isVariableDeclarator(path);
        }}
      >
        {
          ((matches) => {
            return (
              <div>
                {matches.map((path) => path.node.type)}
              </div>
            );
          })
        }
      </Match>
    </ASTProvider>);

    expect(output.toJSON()).toMatchSnapshot(); 
  });

  it('should support callbacks for lateral composition', () => {
    class LodashAndForEach extends React.Component {
      state = {
        lodash: [],
        forEach: false,
      };

      handleImportMatch = (matches) => {
        if (matches.length >= 1) {
          const identifiers = getIdentifiersFromImport(matches);
          this.setState({
            lodash: identifiers,
          });
        }
      };

      handleMemberExpression = (matches) => {
        if (matches.length >= 1) {
          this.setState({
            forEach: true,
          });
        }
      };

      render() {
        return (
          <div>
            <Import from='lodash' onMatch={this.handleImportMatch} />
            <MemberExpression 
              name='forEach' 
              target={this.state.lodash} 
              onMatch={this.handleMemberExpression} 
            />
            { this.state.lodash.length > 0 && this.state.forEach
              ? 'Yes'
              : 'Nope' }
          </div>
        );
      }
    }

    const output = renderer.create(<ASTProvider source={`
      import _ from 'lodash';
      const fruits = [
        'Apples',
        'Oranges',
        'Grapes',
      ];

      _.forEach(fruits, (fruit) => {
        alert(fruit);
      });
    `}>
      <LodashAndForEach/>
    </ASTProvider>);

    expect(output.toJSON()).toMatchSnapshot(); 
  });

  it('should allow match nesting', () => {
    const output = renderer.create(<ASTProvider source={`
      const name = 'Merrick';
      class Person {
        greet() {
          const age = 26;
        }
      }
    `}>

      <Match
        match={(path) => {
          return t.isClassMethod(path);
        }}
      >
        {
          ((matches) => {
            return (
              <div>
                {matches.map((method, key) => (
                  <Match
                    match={(potentialDeclarator) => {
                      return t.isVariableDeclarator(potentialDeclarator);
                    }}
                    key={key}
                    ast={method}
                  >
                    {
                      ((declarators) => {
                        return (
                          <div>
                            {declarators.map((path, index) => (
                              <div key={index}>{path.node.type}</div>
                            ))}
                          </div>
                        );
                      })
                    }
                  </Match>
                ))}
              </div>
            );
          })
        }
      </Match>
    </ASTProvider>);

    expect(output.toJSON()).toMatchSnapshot(); 
  });
});
