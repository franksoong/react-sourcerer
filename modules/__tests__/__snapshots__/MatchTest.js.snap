exports[`Match should allow match nesting 1`] = `
<div>
  <div>
    <div>
      VariableDeclarator
    </div>
  </div>
</div>
`;

exports[`Match should match a class method 1`] = `
<div>
  ClassMethod
</div>
`;

exports[`Match should match a variable 1`] = `
<div>
  VariableDeclarator
</div>
`;

exports[`Match should support callbacks for lateral composition 1`] = `
<div>
  Yes
</div>
`;

exports[`Match should support lateral by not scoping 1`] = `
<div>
  Yes
</div>
`;
